test/functional/feature_addrman_asmap.py

```diff
@@ -0,0 +1,184 @@
+#!/usr/bin/env python3
+# Copyright (c) 2020-2021 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test addrman/checkaddrman/asmap functionality."""
+
+import os
+import shutil
+import struct
+
+from test_framework.messages import ser_uint256, hash256
+from test_framework.p2p import MAGIC_BYTES
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import assert_equal
+
+
+DEFAULT_ASMAP_FILENAME = "ip_asn.map"  # defined in src/init.cpp
+ASMAP = "../../src/test/data/asmap.raw"  # path to unit test skeleton asmap
+VERSION = "fec61fa21a9f46f3b17bdcd660d7f4cd90b966aad3aec593c99b35f0aca15853"
+
+
+def expected_messages(filename):
+    return [
+        f'Opened asmap file "{filename}" (59 bytes) from disk',
+        f"Using asmap version {VERSION} for IP bucketing",
+    ]
+
+
+def serialize_addrman(*, format=1, lowest_compatible=3):
+    new = []
+    tried = []
+    INCOMPATIBILITY_BASE = 32
+    r = MAGIC_BYTES["regtest"]
+    r += struct.pack("B", format)
+    r += struct.pack("B", INCOMPATIBILITY_BASE + lowest_compatible)
+    r += ser_uint256(1)
+    r += struct.pack("i", len(new))
+    r += struct.pack("i", len(tried))
+    ADDRMAN_NEW_BUCKET_COUNT = 1 << 10
+    r += struct.pack("i", ADDRMAN_NEW_BUCKET_COUNT ^ (1 << 30))
+    for _ in range(ADDRMAN_NEW_BUCKET_COUNT):
+        r += struct.pack("i", 0)
+    checksum = hash256(r)
+    r += checksum
+    return r
+
+
+def write_addrman(peers_dat, **kwargs):
+    with open(peers_dat, "wb") as f:
+        f.write(serialize_addrman(**kwargs))
+
+
+class AddrmanAsmapTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 1
+
+    def set_peers_dat(self):
+        """Copy a peers.dat containing tried table entries into the data directory."""
+        peers_data = os.path.join(os.path.dirname(os.path.realpath(__file__)), "data/peers.dat")
```

Comment:
An alternative to adding a binary file would be to hack something together with the `serialize_addrman` function in this file

Reply:
I agree, it would be a logical next step if this is merged but this can do the job for now.

(For now I think it's fundamental to have these regression tests in place quickly, given the number of recent bugs that escaped review, the reproducible criticality of the last bug that this adds a test for, and the addrman/asmap refactoring taking place.)

Reply:
Obviously not a blocker, but if the file is removed after this is merged, then it would be better to not commit it in the first place and use `serialize_addrman` (or similar) from the beginning.

Reply:
+1 to approach of `serialize_addrman` function.

Reply:
Happy to look at it as a follow-up (or leave it for someone), but this does the job and by common sense ought to have been the first step in a responsible fix (or at least part of it), rather than ignored after https://github.com/bitcoin/bitcoin/pull/22791#issuecomment-907856179 while embarking on many refactoring and changes, no matter how desirable they may have been. Foot-dragging and bikeshedding doesn't seem appropriate here compared to the gravity of the bug that is covered.

Reply:
> Happy to look at it as a follow-up (or leave it for someone)

Happy to propose an alternative that uses `serialize_addrman`, if you don't want to do it here. I just think that we shouldn't be adding binary files to the git repo, especially short-lived ones, when it can be avoided. Personally I'd like to see both approaches, so that they can be compared and the best one be picked. Though, again, not a blocker. If this gets review, I am happy to merge nonetheless.

> while embarking on many refactoring

I am not aware of any refactoring changes that have been merged in the meantime after the bug was discovered.

> gravity of the bug that is covered

I agree that test coverage for that is badly needed. However, the existence of a regression test doesn't mean it should be merged without review. (At least I couldn't find any review ACKs on this pull).

Reply:
> I am not aware of any refactoring changes that have been merged in the meantime after the bug was discovered.

Instead of a straight revert, or simple reordering, #22791 was a larger than necessary diff with no accompanying tests.